### PR TITLE
API: Change GetProjectedIds to Return all Ids

### PR DESCRIPTION
**Pull request by RussellSpitzer**
Previously getProjectedIds would only return leaf nodes and primitives that
were selected. This made it impossible to return empty structs. To fix this
we change the behavior to return all id's of required fields including structs.
This in turn requires fixing the alternate PruneColumn methods for Avro
and Parquet to respect that they will now have selected field ID's for non
primitive nodes. Previous use cases of TypeUtil.select are converted to
TypeUtil.project, which inverses this new getProjecetedIds code.

### CODE REVIEW

File: api/src/main/java/org/apache/iceberg/types/GetProjectedIds.java [hunk 1]
```diff
@@ -39,9 +39,7 @@
 
   @Override
   public Set<Integer> field(Types.NestedField field, Set<Integer> fieldResult) {
-    if (fieldResult == null) {
-      fieldIds.add(field.fieldId());
-    }
+    fieldIds.add(field.fieldId());
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/types/GetProjectedIds.java:42]: This is related to [my comment on the other PR](https://github.com/apache/iceberg/pull/2952/files#r685370873). I don't think that we should return inner IDs other than struct IDs or else it isn't clear how lists and maps should be handled.

**Comment by kbendick** [comment]
> This in turn requires fixing the alternate PruneColumn methods for Avro
> and Parquet to respect that they will now have selected field ID's for non
> primitive nodes.

Do you know if anything needs to be done for ORC @RussellSpitzer? I'm helping out with ORC more going forward and if you're aware of anything that needs to be updated there, if you don't have time to update it, if you make an issue I'll see if I can grab it (or at the least we'll have the issue to track).

If you're not sure then possibly you can add a unit test for ORC as well? If it fails, open an issue (or I will) and then we can follow up on it after. =)

File: api/src/main/java/org/apache/iceberg/types/TypeUtil.java [hunk 2]
```diff
@@ -130,13 +130,13 @@ public static Schema select(Schema schema, Set<Integer> fieldIds) {
   public static Types.StructType selectNot(Types.StructType struct, Set<Integer> fieldIds) {
     Set<Integer> projectedIds = getIdsInternal(struct);
     projectedIds.removeAll(fieldIds);
-    return select(struct, projectedIds);
+    return project(struct, projectedIds);
```
Comment (RussellSpitzer) [inline api/src/main/java/org/apache/iceberg/types/TypeUtil.java:133]: One issue here is selectNot doest not actually deselect children when a parent ID is not selected. Previously this is because getProjectedIDs (behind getIdsInternal) would not return parent struct ids, so removing it from the set of projectedIds would not do anything.

Now It will not work because removing a parentID still leaves all child IDs. We could fix this but it would be a change in behavior from the previous code.

**Comment by RussellSpitzer** [comment]
> > This in turn requires fixing the alternate PruneColumn methods for Avro
> > and Parquet to respect that they will now have selected field ID's for non
> > primitive nodes.
> 
> Do you know if anything needs to be done for ORC @RussellSpitzer? I'm helping out with ORC more going forward and if you're aware of anything that needs to be updated there, if you don't have time to update it, if you make an issue I'll see if I can grab it (or at the least we'll have the issue to track).
> 
> If you're not sure then possibly you can add a unit test for ORC as well? If it fails, open an issue (or I will) and then we can follow up on it after. =)

I don't think so, we don't have any custom ORC projection code as far as I can tell. It just uses the output of a TypeUtil.selectNot() 

https://github.com/apache/iceberg/blob/68091037944ff7e9de91e7b619f313a8e98c1adc/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java#L158-L159

against the ORC schema to determine what to read and then uses that directly here : 

https://github.com/apache/iceberg/blob/970e8aac60081c7d286ab0f7e12e428b3309dd34/orc/src/main/java/org/apache/iceberg/orc/ORC.java#L315-L319

File: api/src/main/java/org/apache/iceberg/types/TypeUtil.java (hunk 2 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/types/TypeUtil.java:135]: I think I agree with the decision to not change the behavior of this method, even though the opposite of "select" behavior would be to fully remove a struct when its ID is passed in `fieldIds`.

But I don't think that `project` is quite correct either. Consider the example schema `1: id bigint, 2: location struct<3: lat double, 4: long double>`. Previously, `selectNot(t, set(3, 4))` would produce `1: id bigint` and omit the location entirely. Using project with the updated `GetProjectedIds`, the projected ID set will be {1, 2, 3, 4} and not `{1, 3, 4}`. That would result in the same call producing `1: id bigint, 2: location struct<>`, which introduces a new bug because now there is an unexpected extra field.

To clean this up, I think we need a version of `GetProjectedIds` that doesn't select structs and uses the old behavior.

File: api/src/main/java/org/apache/iceberg/util/StructProjection.java [hunk 3]
```diff
@@ -42,7 +42,7 @@
    */
   public static StructProjection create(Schema schema, Set<Integer> ids) {
     StructType structType = schema.asStruct();
-    return new StructProjection(structType, TypeUtil.select(structType, ids));
+    return new StructProjection(structType, TypeUtil.project(structType, ids));
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/util/StructProjection.java:45]: Looks like there aren't any uses of this call, which is good. I agree that we probably want this to use `project` instead of `select`.

File: core/src/main/java/org/apache/iceberg/BaseTableScan.java [hunk 4]
```diff
@@ -296,7 +296,7 @@ private Schema lazyColumnProjection() {
       }
       requiredFieldIds.addAll(selectedIds);
 
-      return TypeUtil.select(schema, requiredFieldIds);
+      return TypeUtil.project(schema, requiredFieldIds);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTableScan.java:299]: I agree with this because it is the opposite of `GetProjectedIds` used above.

File: core/src/main/java/org/apache/iceberg/avro/PruneColumns.java [hunk 5]
```diff
@@ -81,13 +81,18 @@ public Schema record(Schema record, List<String> names, List<Schema> fields) {
...
+      // types can be selected by projecting the keys, values, or elements. Empty
+      // Structs can be selected by selecting the record itself instead of it's children.
       // This creates two conditions where the field should be selected: if the
       // id is selected or if the result of the field is non-null. The only
       // case where the converted field is non-null is when a map or list is
       // selected by lower IDs.
       if (selectedIds.contains(fieldId)) {
-        filteredFields.add(copyField(field, field.schema(), fieldId));
+        if (fieldSchema != null) {
+          filteredFields.add(copyField(field, fieldSchema, fieldId));
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:92]: I'm not sure that I understand the reason for this change. Is this implementing the same change as the previous PR, but in the Avro `PruneColumns`?

It looks like if a struct field is selected and a sub-field is selected, then the selection for the struct isn't a full selection. But if a sub-field is not selected then the selection for the struct is a full selection. That doesn't make sense to me.

File: core/src/test/java/org/apache/iceberg/TestSchemaUpdate.java [hunk 6]
```diff
@@ -93,7 +93,7 @@ public void testDeleteFields() {
       Schema del = new SchemaUpdate(SCHEMA, 19).deleteColumn(name).apply();
 
       Assert.assertEquals("Should match projection with '" + name + "' removed",
-          TypeUtil.select(SCHEMA, selected).asStruct(), del.asStruct());
+          TypeUtil.project(SCHEMA, selected).asStruct(), del.asStruct());
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestSchemaUpdate.java:96]: Looks correct.

File: core/src/test/java/org/apache/iceberg/avro/TestReadProjection.java [hunk 7]
```diff
@@ -37,10 +37,11 @@
...
 
 public abstract class TestReadProjection {
+
   protected abstract Record writeAndRead(String desc,
-                                         Schema writeSchema,
-                                         Schema readSchema,
-                                         Record record) throws IOException;
+      Schema writeSchema,
+      Schema readSchema,
+      Record record) throws IOException;
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/avro/TestReadProjection.java:44]: Can you revert the non-functional changes here? This one and the whitespace one above.

File: spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 8]
```diff
@@ -1034,7 +1061,7 @@ public void testUnpartitionedPartitionsTable() {
   @Test
   public void testPartitionsTable() {
     TableIdentifier tableIdentifier = TableIdentifier.of("db", "partitions_test");
-    Table table = createTable(tableIdentifier, SCHEMA, PartitionSpec.builderFor(SCHEMA).identity("id").build());
+    Table table = createTable(tableIdentifier, SCHEMA, SPEC);
```
Comment (rdblue) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1064]: Are these changes needed? I don't have a problem with them, but I don't think they are required right?

File: core/src/main/java/org/apache/iceberg/avro/PruneColumns.java (hunk 5 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:92]: As I'm thinking about this more, I think that the behavior in this class should always match `project`. I doubt there's a case where we want select behavior, right? In that case, shouldn't the `else` case check whether the type is a record and create an empty record?

File: api/src/main/java/org/apache/iceberg/types/TypeUtil.java (hunk 2 above)
Comment (RussellSpitzer) [reply api/src/main/java/org/apache/iceberg/types/TypeUtil.java:135]: That seems like the right behavior to me? Shouldn't you be required to explicitly omit the parent if you don't want the that element? Otherwise there would be no way to "selectNot"  and only get back the empty struct.

Comment (RussellSpitzer) [reply api/src/main/java/org/apache/iceberg/types/TypeUtil.java:135]: Wrote up these test cases, i'll run the full test suite to make sure this works with our other usages

```
    Schema schema = new Schema(
        Lists.newArrayList(
            required(1, "id", Types.LongType.get()),
            required(2, "location", Types.StructType.of(
                required(3, "lat", Types.DoubleType.get()),
                required(4, "long", Types.DoubleType.get())
            ))));

    Schema expectedNoPrimitive = new Schema(
        Lists.newArrayList(
            required(2, "location", Types.StructType.of(
                required(3, "lat", Types.DoubleType.get()),
                required(4, "long", Types.DoubleType.get())
            ))));

    Schema actualNoPrimitve = TypeUtil.selectNot(schema, Sets.newHashSet(1));
    Assert.assertEquals(expectedNoPrimitive.asStruct(), actualNoPrimitve.asStruct());

    // Expected legacy behavior is to completely remove structs if their elements are removed
    Schema expectedNoStructElements = new Schema(required(1, "id", Types.LongType.get()));
    Schema actualNoStructElements = TypeUtil.selectNot(schema, Sets.newHashSet(3, 4));
    Assert.assertEquals(expectedNoStructElements.asStruct(), actualNoStructElements.asStruct());

    // Expected legacy behavior is to ignore selectNot on struct elements.
    Schema actualNoStruct = TypeUtil.selectNot(schema, Sets.newHashSet(2));
    Assert.assertEquals(schema.asStruct(), actualNoStruct.asStruct());
    ```

File: spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java (hunk 8 above)
Comment (RussellSpitzer) [reply spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1064]: No, just a clean up I had in the original PR, I can split this into another pull request if needed

**Comment by RussellSpitzer** [comment]
@rdblue All fixed up

File: parquet/src/main/java/org/apache/iceberg/parquet/PruneColumns.java [hunk 9]
```diff
@@ -49,7 +49,12 @@ public Type message(MessageType message, List<Type> fields) {
       Type field = fields.get(i);
       Integer fieldId = getId(originalField);
       if (fieldId != null && selectedIds.contains(fieldId)) {
-        builder.addField(originalField);
+        if (field != null) {
+          hasChange = true;
+          builder.addField(field);
+        } else {
+          builder.addField(originalField);
```
Comment (RussellSpitzer) [inline parquet/src/main/java/org/apache/iceberg/parquet/PruneColumns.java:56]: Should I do the Record Message only thing here as well? where we copy the struct to be empty?

File: api/src/main/java/org/apache/iceberg/types/TypeUtil.java [hunk 10]
```diff
@@ -125,20 +125,24 @@ public static Schema select(Schema schema, Set<Integer> fieldIds) {
     return ImmutableSet.copyOf(getIdsInternal(type));
   }
 
+  private static Set<Integer> getIdsInternal(Type type, boolean includeStructIds) {
+    return visit(type, new GetProjectedIds(includeStructIds));
+  }
+
   private static Set<Integer> getIdsInternal(Type type) {
-    return visit(type, new GetProjectedIds());
+    return getIdsInternal(type, true);
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/types/TypeUtil.java:133]: Is it a lot of changes to remove this method? I wouldn't expect it since this is `private`. I'd probably prefer removing it, but up to you if it touches a bunch of unrelated code.

File: core/src/main/java/org/apache/iceberg/avro/PruneColumns.java [hunk 11]
```diff
@@ -81,13 +83,22 @@ public Schema record(Schema record, List<String> names, List<Schema> fields) {
...
+        } else {
+          if (isRecord(field.schema())) {
+            filteredFields.add(copyField(field, makeEmptyCopy(field.schema()), fieldId));
+          } else {
+            filteredFields.add(copyField(field, field.schema(), fieldId));
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:99]: Okay, so in this case the field is a map, list, or primitive. Then we just follow the old behavior. Looks good to me since the Iceberg schema selection would fail.

File: core/src/main/java/org/apache/iceberg/avro/PruneColumns.java [hunk 12]
```diff
@@ -81,13 +83,22 @@ public Schema record(Schema record, List<String> names, List<Schema> fields) {
...
+      // types can be selected by projecting the keys, values, or elements. Empty
+      // Structs can be selected by selecting the record itself instead of its children.
       // This creates two conditions where the field should be selected: if the
       // id is selected or if the result of the field is non-null. The only
       // case where the converted field is non-null is when a map or list is
       // selected by lower IDs.
       if (selectedIds.contains(fieldId)) {
-        filteredFields.add(copyField(field, field.schema(), fieldId));
+        if (fieldSchema != null) {
+          filteredFields.add(copyField(field, fieldSchema, fieldId));
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:94]: I think we need to set `hasChange` in the cases where we don't return `field.schema()` for the field, right?

File: parquet/src/main/java/org/apache/iceberg/parquet/PruneColumns.java [hunk 13]
```diff
@@ -49,7 +51,16 @@ public Type message(MessageType message, List<Type> fields) {
       Type field = fields.get(i);
       Integer fieldId = getId(originalField);
       if (fieldId != null && selectedIds.contains(fieldId)) {
-        builder.addField(originalField);
+        if (field != null) {
+          hasChange = true;
+          builder.addField(field);
+        } else {
+          if (isStruct(originalField)) {
+            builder.addField(originalField.asGroupType().withNewFields(Collections.emptyList()));
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/PruneColumns.java:59]: Should `hasChange` be true here as well?

File: parquet/src/main/java/org/apache/iceberg/parquet/PruneColumns.java (hunk 9 above)
Comment (rdblue) [reply parquet/src/main/java/org/apache/iceberg/parquet/PruneColumns.java]: What do you mean?

**Comment by rdblue** [comment]
@RussellSpitzer, looking good! My one remaining concern is the `hasChange` behavior. That may also be causing tests to not do what you expect.

File: api/src/main/java/org/apache/iceberg/types/TypeUtil.java (hunk 10 above)
Comment (RussellSpitzer) [reply api/src/main/java/org/apache/iceberg/types/TypeUtil.java:133]: We are pretty safe here, removing it. Only two changes in this file

File: core/src/main/java/org/apache/iceberg/avro/PruneColumns.java (hunk 12 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:94]: I think we are actually fine here unless every field is selected because the logic for has change is a bit confusing.

You either 

1. Have a change (Make a new record using the filtered fields)
2. Have no change and filtered fields size is the same as the original number of fields
3. Have no change and filtered field size is not empty (Make a new record using the filtered fields)

Currently we have tests hitting 1 and 3 but not 2 :/ 
I'll add the "hasChange" flag

File: parquet/src/main/java/org/apache/iceberg/parquet/PruneColumns.java (hunk 13 above)
Comment (RussellSpitzer) [reply parquet/src/main/java/org/apache/iceberg/parquet/PruneColumns.java:59]: Changing the ordering of the statements under (field != null) so that they match the usages here

**Comment by RussellSpitzer** [comment]
@rdblue patched up, I think I got that covered. Should be set now. I also removed that private function as recommended

File: core/src/main/java/org/apache/iceberg/avro/PruneColumns.java (hunk 12 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:95]: This looks good now.

File: core/src/test/java/org/apache/iceberg/avro/TestReadProjection.java [hunk 14]
```diff
@@ -526,4 +526,183 @@ public void testListOfStructsProjection() throws IOException {
...
+            Types.NestedField.required(1, "lat", Types.FloatType.get()),
+            Types.NestedField.required(2, "long", Types.FloatType.get())
+        ))
+    );
+
+    Record record = new Record(AvroSchemaUtil.convert(writeSchema, "table"));
+    record.put("id", 34L);
+    Record location = new Record(record.getSchema().getField("location").schema());
+    location.put("lat", 52.995143f);
+    location.put("long", -1.539054f);
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/avro/TestReadProjection.java:576]: Odd [location](https://www.google.com/maps/@52.995143,-1.539054,13z) to choose.

File: core/src/test/java/org/apache/iceberg/avro/TestReadProjection.java [hunk 15]
```diff
@@ -526,4 +526,183 @@ public void testListOfStructsProjection() throws IOException {
...
+    Record result = (Record) projected.get("location");
+
+    Assert.assertEquals("location should be in the 0th position", result, projected.get(0));
+    Assert.assertNotNull("Should contain an empty record", result);
+    AssertHelpers.assertEmptyAvroField(result, "lat");
+    AssertHelpers.assertEmptyAvroField(result, "long");
+  }
+
+  @Test
+  public void testEmptyStructRequiredProjection() throws Exception {
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/avro/TestReadProjection.java:563]: Isn't this identical to the test case above?

Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/avro/TestReadProjection.java:563]: Nevermind, I see that the write schema has the struct as optional.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @RussellSpitzer!

File: core/src/test/java/org/apache/iceberg/avro/TestReadProjection.java (hunk 14 above)
Comment (RussellSpitzer) [reply core/src/test/java/org/apache/iceberg/avro/TestReadProjection.java:576]: It was already in the test suite and came in the Netflix original commit :) So you'll have to ask whoever wrote the first version.

https://github.com/apache/iceberg/blame/master/core/src/test/java/org/apache/iceberg/avro/TestReadProjection.java#L205-L206

**Comment by RussellSpitzer** [comment]
Thanks @rdblue I know this one was a bit of a pain, but we should hopefully have all of our projection issues fixed now 🤞

**Comment by szehon-ho** [comment]
🎉 Great work !
